File: 146-lru-cache/lru-cache.cpp


class LRUCache {
private:
    int capacity;
    list<pair<int, int>> items; // Doubly-linked list to store key-value pairs in access order.
    unordered_map<int, list<pair<int, int>>::iterator> cache; // Hash map to store key and iterator to the corresponding item in the list.

public:
    LRUCache(int capacity) : capacity(capacity) {}

    int get(int key) {
        auto foundIt = cache.find(key);
        if (foundIt == cache.end()) {
            return -1; // Key not found.
        }
        // Move the accessed item to the front of the list.
        items.splice(items.begin(), items, foundIt->second);
        return foundIt->second->second;
    }

    void put(int key, int value) {
        auto foundIt = cache.find(key);
        if (foundIt != cache.end()) {
            // Update existing item's value and move it to the front of the list.
            foundIt->second->second = value;
            items.splice(items.begin(), items, foundIt->second);
            return;
        }
        // Check if the cache is at capacity.
        if (items.size() == capacity) {
            // Remove the least recently used item from both the list and the hash map.
            auto lru = items.back();
            cache.erase(lru.first);
            items.pop_back();
        }
        // Insert the new item at the front of the list.
        items.emplace_front(key, value);
        // Update the hash map with the new item's iterator.
        cache[key] = items.begin();
    }
};

File: 146-lru-cache/lru-cache.py
class Node:
    def __init__(self, key, value):
        self.key = key        # store the key (needed for removal from hashmap)
        self.value = value    # store the value
        self.prev = None      # pointer to previous node
        self.next = None      # pointer to next node

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = {}  # maps key -> node

        # Dummy head and tail to avoid edge checks
        self.head = Node(0, 0)  # least recently used
        self.tail = Node(0, 0)  # most recently used
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node):
        """Remove node from the linked list."""
        prev = node.prev
        nxt = node.next
        prev.next = nxt
        nxt.prev = prev

    def _add_to_tail(self, node):
        """Add node right before the tail (most recently used)."""
        prev = self.tail.prev
        prev.next = node
        node.prev = prev
        node.next = self.tail
        self.tail.prev = node

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.cache:
            node = self.cache[key]
            self._remove(node)
            self._add_to_tail(node)
            return node.value
        return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.cache:
            self._remove(self.cache[key])
        
        node = Node(key, value)
        self.cache[key] = node
        self._add_to_tail(node)

        if len(self.cache) > self.capacity:
            # Remove LRU from head
            lru = self.head.next
            self._remove(lru)
            del self.cache[lru.key]
